IntegrationTests/UITests/UITest.cs

```diff
@@ -1,58 +1,181 @@
 ﻿using System;
+using System.Diagnostics;
 using System.Linq;
+using System.Reflection;
 using System.Threading.Tasks;
 using System.Windows.Forms;
+using CommonTestUtils;
 using GitUI;
+using NUnit.Framework;
 
 namespace GitExtensions.UITests
 {
     public static class UITest
     {
-        public static async Task WaitForIdleAsync()
+        public static void RunForm<T>(
+            Action showForm,
+            Func<T, Task> runTestAsync)
+            where T : Form
         {
+            LoggingService.Log($"{MethodBase.GetCurrentMethod().Name} entry");
             var idleCompletionSource = new TaskCompletionSource<VoidResult>();
+
+            // Start runAsync before calling showForm.
+            // The latter might block until the form is closed, especially if using Application.Run(form).
+
+            // Avoid using ThreadHelper.JoinableTaskFactory for the outermost operation because we don't want the task
+            // tracked by its collection. Otherwise, test code would not be able to wait for pending operations to
+            // complete.
+            var test = ThreadHelper.JoinableTaskContext.Factory.RunAsync(async () =>
+            {
+                LoggingService.Log($"{MethodBase.GetCurrentMethod().Name}.test entry");
+
+                // Wait for the form to be opened by the test thread.
+                await idleCompletionSource.Task;
+                LoggingService.Log($"{MethodBase.GetCurrentMethod().Name}.test idle");
+                var form = Application.OpenForms.OfType<T>().Single();
+                if (!form.Visible)
```

Comment:
📝 Switching to the main thread needs to occur before these APIs are used.

Reply:
I will change this.

This pointed me to another -- maybe the actual -- race condition:
The main thread which runs the form might even not exist yet when the async task is started,
because `Application.Run(form)` will first create it.
Then `await SwitchToMainThreadAsync` should have thrown "no main thread" but it seems to block.